Expose SPIR-V information in Vulkan.SPIRVInfo

expipiplus1 opened the pull request:
Closes https://github.com/expipiplus1/vulkan/issues/214

expipiplus1 commented:
@sheaf: https://github.com/expipiplus1/vulkan/blob/176b103ef3d0d913acf2f9560b60d62fd949feb0/src/Vulkan/SPIRVInfo.hs

Types are a bit basic at the moment, but that's all quite easy to change.

The extension names are useful at runtime, but the Features and Caps are just sad as they are.

They should really be functions which modify a `SomeStruct DeviceFeatures2` and insert the required feature, or (check it). That may be a little cumbersome to implement though

sheaf commented:
This is excellent.    

Would it be possible to add record fields to the datatype? So that we have:    

```haskell
data SPIRVRequirement
  = Version { version :: Version }
  | Extension { extension :: ByteString }
  | Feature { struct :: ByteString, feature :: ByteString, requires :: Vector ByteString }
  | Property { property :: ByteString, member :: ByteString, value :: ByteString, requires :: Vector ByteString }
```

I know record fields in a sum are partial (which is a shame), but it helps to distinguish between fields that otherwise have the same type. I much prefer writing

```haskell
f :: SPIRVRequirement -> ...
...
f ( Feature { struct, feature, requires } ) = ...
f ( Property { property, member, value, requires } ) = ...
```

than

```haskell
g :: SPIRVRequirement -> ...
...
g ( Feature struct feature requires ) = ...
g ( Property property member value requires ) = ...
```

Otherwise, I'm very happy with this, thanks a lot.

expipiplus1 commented:
> This is excellent.

Thanks :)

> Would it be possible to add record fields to the datatype? So that we have:

Yeah, shouldn't be a problem.

An alternative might be having four functions for extensions and four for capabilities, i.e. one function per version/extension/feature/property/. After all, the user's only going to have to unpack the result immediately anyway...

(the downside is doing the lookup 4 times, but eh, it's not something that's going to happen in a tight loop)

sheaf commented:
I'm trying to think through how it would work if we wanted to properly consume all this info, given a collection of SPIR-V extensions and capabilities.


I'm thinking of using a GADT:

```haskell
data EnableStruct
  = NoStruct
  | FeatureStruct Type
  | PropertyStruct Type

data Enable ( p :: EnableStruct ) where
  Version :: { version :: Version } -> Enable 'NoStruct
  Extension :: { extension :: Extension } -> Enable 'NoStruct
  Feature 
    :: forall struct
    .  ( PokeStruct struct, Extends DeviceCreateInfo struct )
    => { checkFeature  :: ( forall fs. Elem struct fs => PhysicalDeviceFeatures2 fs -> Bool )
       , enableFeature :: struct -> struct
       , mbMinVersion  :: Maybe Version
       , enableExts    :: [ Extension ]
       }
    -> Enable ( 'FeatureStruct struct )
  Property
    :: forall struct
    .  ( Extends PhysicalDeviceProperties2 struct )
    => { checkProperty :: ( forall ps. Elem struct ps => PhysicalDeviceProperties2 ps -> Bool )
       , mbMinVersion  :: Maybe Version
       , enableExts    :: [ Extension ]
       }
    -> Enable ( 'PropertyStruct struct )
```

Then we would be able to obtain the necessary "features" chain and "properties" chain at the type level, use that to query the device for the appropriate features/properties, and then pass them to the function that creates the logical device.

sheaf commented:
OK, I've written a module that takes a bunch of requirements and figures out the corresponding device feature and property structure chains. Please take a look [here](https://github.com/sheaf/vulkan-1/commit/75381289953f1a14643065541e1de2750933385f).

expipiplus1 commented:
This looks great, thanks!

A couple of things:

- Some features aren't mentioned in the tail of the struct chain, for instance the SPIR-V capability `Geometry` requires the `geometryShader` feature in [`PhysicalDeviceFeatures`](https://hackage.haskell.org/package/vulkan-3.6.15/docs/Vulkan-Core10-DeviceInitialization.html#t:PhysicalDeviceFeatures) 
- Some features can be enabled in more than one struct type in the chain, for example the SPIR-V capability `MultiView` requires the `multiview` feature which can be enabled using either [`PhysicalDeviceVulkan11Features`](https://hackage.haskell.org/package/vulkan-3.6.15/docs/Vulkan-Core12.html#t:PhysicalDeviceVulkan11Features) or [`PhysicalDeviceMultiviewFeatures`](https://hackage.haskell.org/package/vulkan-3.6.15/docs/Vulkan-Core11-Promoted_From_VK_KHR_multiview.html#t:PhysicalDeviceMultiviewFeatures)

I've not thought about a solution much, but perhaps requirements should be represented not as a single `Requirement` but as a disjunction of `Requirement`s, and instead of focusing on a particular struct in the chain be able to examine the chain as a whole.

expipiplus1 commented:
Also I suspect that `Requirements` will need to cart around a `PokeChain feats` and `props` to actually be useful which is disappointingly redundant with `KnownChain`. Perhaps we should remove `PokeChain` and `PeekChain` and just have `KnownChain`, certainly seems nicer...

sheaf commented:
I was mostly going off `vk.xml`, where the requirements from `SPIR-V` capabilities/extensions all needed something from `PhysicalDeviceFeatures2`/`PhysicalDeviceProperties2` and all used `PhysicalDeviceVulkan11Features`/`PhysicalDeviceVulkan12Features` when possible (instead of individual feature structures). For instance, the `MultiView` capability seems to specifically ask to be set in `VkPhysicalDeviceVulkan11Features` and not `PhysicalDeviceMultiviewFeatures`:

```xml
<spirvcapability name="MultiView">
  <enable struct="VkPhysicalDeviceVulkan11Features" feature="multiview" requires="VK_VERSION_1_1,VK_KHR_multiview"/>
```

I agree that if we want to use this to specify any features/properties (not just those associated to `SPIR-V` modules), it should also support base features/properties.     
For the issue with using `PhysicalDeviceVulkanXxFeatures` or individual feature structs, this can be decided using the `Vulkan` API version. I am assuming that no structures are incompatible (in the sense that they can't both be present in the `PhysicalDeviceFeatures2`/`PhysicalDeviceProperties2` structure chains) other than this specific situation (at least, that's what I'm seeing [here](https://hackage.haskell.org/package/vulkan-3.6.15/docs/Vulkan-Core10-Device.html#t:DeviceCreateInfo)).    

For me, the whole point is that this allows users to specify any collection of requirements and have it generate the necessary structure chain from it, as opposed to asking users to come up with this instance chain themselves which can be tedious.

expipiplus1 commented:
> specifically ask to be set in `VkPhysicalDeviceVulkan11Features` and not `PhysicalDeviceMultiviewFeatures`

I think this is a deficiency in the spec, the requirements `VK_VERSION_1_1,VK_KHR_multiview` mean either 1.1 or KHR_multiview need to be enabled, and one can't use the specified struct `VkPhysicalDeviceVulkan11Features` with the latter.

(I don't think that this is really a big issue though, so happy with any solution)

expipiplus1 commented:
> For me, the whole point is that this allows users to specify any collection of requirements and have it generate the necessary structure chain from it, as opposed to asking users to come up with this instance chain themselves (which can be tedious).

:+1:

sheaf commented:
[New version](https://github.com/sheaf/vulkan-1/commit/229b89b280e8cb0605ab59ae74e3e8cef7d5c511) which allows specification of base features/properties (i.e. those found in `PhysicalDeviceFeatures` and `PhysicalDeviceProperties`).

expipiplus1 commented:
Cool, I've added it at `src-manual` (as `src` is liable to be overwritten by the generator)

SPIR-V requirements are also implemented: https://github.com/expipiplus1/vulkan/blob/ecd19c3cdf2b4ebe75ec9ad2292b4f05e70bdbc0/src/Vulkan/SPIRVRequirements.hs

it compiles, but I've not tested it yet.

expipiplus1 commented on generate-new/vk/VK/SPIRVElements.hs:
```diff
@@ -0,0 +1,280 @@
+renderSPIRVThing funName name reqs xs = do
+  let case' x = do
+        let mergeReqs = Prelude.head -- TODO: this is probably wrong! discarding the
+                                     -- other reqs
```
This is not correct! It's only paying attention to the first "enable" node for each capability/requirement

See here for an example: https://github.com/KhronosGroup/Vulkan-Docs/blob/9f9fd160c8dbd8752447a4a8a2e32fbb92b9722d/xml/vk.xml#L14907-L14910

expipiplus1 replied:
I think the best thing to do would be to also take a version, and if it is at least a version mentioned in the "enable" then that enable is returned, otherwise the other one is.

expipiplus1 commented:
I've incorporated your changes into this branch with some (quite major I'm afraid) modifications.

Don't have time to write about it now, but will be around tomorrow.

This is the change to using them in examples: https://github.com/expipiplus1/vulkan/pull/216/commits/c68d9ae6812b9895226bd47688b3bf9cee2d76c3

They're a little cumbersome at the moment, but I'm sure that can be cleared up

sheaf commented on utils/src/Vulkan/Utils/Requirements.hs:
```diff
@@ -0,0 +1,554 @@
+-- * Instance Creation
+----------------------------------------------------------------
+
+checkInstanceRequirements
```
This function doesn't check that the provided `InstanceCreateInfos es` is allowable, i.e. that all the extension structures are compatible with the instance. That's fine, I think it's reasonable to assume `InstanceCreateInfos es` is valid relative to what is returned by `enumerateInstanceExtensionProperties`, but at the moment (as far as I can see there), is no mechanism that would create a structure chain corresponding to a set of extensions that the user asks for (`makeInstanceCreateInfo` doesn't do anything about extension structures, it keeps the same chain). This was what the (now removed) `InstanceSetting` constructor was used for. I envisioned that a `SPIR-V` extension might well specifically ask for a field in an instance extension structure to be set, which is why I included it in `Requirement`.

expipiplus1 replied:
Ah, yeah, I was considering that it was much easier for the user to just `:&` on the extension structures (instead of carrying around all sorts of peek/poke/extends proofs in the InstanceRequirement GADT)

At the moment it's only these four:

https://github.com/expipiplus1/vulkan/blob/0a59ebc9ce87f6b60dca93afc122f1c78abe4e34/src/Vulkan/CStruct/Extends.hs#L872-L875

sheaf commented:
OK, so you're using a `DMap TypeRep` instead of the existential `Requirements` GADT, that seems to be a good idea.    

The dependency stuff is also neat, glad to have that.

Left a few comments inline. Mainly:
- I don't think it is safe to call  `getPhysicalDeviceProperties`/`getPhysicalDeviceFeatures` unless we know we know we are using Vulkan 1.1 at least,
- it seems you removed the ability to automatically generate a structure chain of instance extensions, but I agree that the use case is probably not very compelling currently so fair enough.

expipiplus1 commented on utils/src/Vulkan/Utils/Requirements.hs:
```diff
@@ -0,0 +1,554 @@
+class (KnownChain es, Extendss PhysicalDeviceProperties2 es) => DevicePropertyChain es where
+instance (KnownChain es, Extendss PhysicalDeviceProperties2 es) => DevicePropertyChain es where
+
+type ChainCont h c a =
```
Thanks, corrected

expipiplus1 commented on utils/src/Vulkan/Utils/Requirements.hs:
```diff
@@ -0,0 +1,554 @@
+  -> o InstanceRequirement
+  -- ^ Optional requests
+  -> InstanceCreateInfo es
+  -- ^ An 'InstanceCreateInfo' with no extensions. If you need elements in the
```
Thanks

expipiplus1 commented on utils/src/Vulkan/Utils/Requirements.hs:
```diff
@@ -0,0 +1,554 @@
+      -- Fetch everything
+      --
+      feats           <- getPhysicalDeviceFeatures2 @fs phys
+      props           <- getPhysicalDeviceProperties2 @ps phys
```
Oh yeah, good point

expipiplus1 replied:
Fixed, thanks,

expipiplus1 commented on utils/src/Vulkan/Utils/Requirements.hs:
```diff
@@ -0,0 +1,554 @@
+       -> ByteString
+       -> Maybe ExtensionProperties
+       )
+getLookupExtension mbPhys deviceExtensionLayers = do
```
:+1:

expipiplus1 commented:
> They're a little cumbersome at the moment, but I'm sure that can be cleared up

Check out https://github.com/expipiplus1/vulkan/pull/216/commits/355366a5ede91261de0b02efa05a16c660598fa1

sheaf commented:
> Check out [355366a](https://github.com/expipiplus1/vulkan/commit/355366a5ede91261de0b02efa05a16c660598fa1)

Neat!
